storyruntime/processing/Lexicon.py

```
@@ -329,59 +329,91 @@ def unless_condition(logger, story, line):
-        :return: If a finally block is present, returns the line number of the finally block, otherwise, the line to be executed immediately after the catch block."""
+        """
+        Executes the try/catch/finally construct. If any StoryscriptError
+        exception is thrown by the try block, the catch block will be
+        invoked. However, if the error is not of type StoryscriptError,
+        then it will be thrown up directly - in this case, the finally
+        block will not be executed either (since the error that
+        occurred is not a StoryscriptError, but rather a programming
+        error in the runtime).
+
+        :return: Returns the line to be executed immediately after
+        the catch block or finally block.
+        """
         next_line = story.next_block(line)
 
         if next_line is None:
             return None
 
+        async def next_block():
+            """
+            This will execute if the next block is a finally block.
+            It happens because the lexicon should always execute
+            a finally block when there's a StoryscriptError
+
+            :return: Returns the next line to be executed.
+            """
+            if next_line['method'] != 'finally':
+                last_block = story.next_block(next_line)
+            else:
+                last_block = next_line
+
+            if last_block is not None and \
+                    last_block['method'] == 'finally':
+                await Lexicon.execute_block(logger, story,
+                                            last_block)
+                last_block = story.next_block(last_block)
+
+            return Lexicon.line_number_or_none(last_block)
+
+        def clean_output(_output, original_value):
+            if _output is not None and \
+                    _output in story.context:
+                del story.context[_output]
+                if original_value is not None:
+                    story.context[_output] = original_value
```

This looks suspicious like the wrong thing to do. You need to implement the concept of scopes and stack frames anyhow for other stuff.
stack frame:
- no access to parent scope (e.g. when a function gets called)
- save old context in a stack of stack frame and pop the current context when returning back from the function call

scope:
- have access to the parent scope
- BUT: variables written in one scope only last in that scope

---

There already is an issue for this: https://github.com/storyscript/runtime/issues/329

---

Is there a reason why we would not want to define variables to not be accessible after execution? Generally, a catch block would be able to define, and access globals variables. I'm just unsure, maybe I should move the discussion into #329 

In this specific case, I actually added the code due to a recent error I discovered in the compiler. The catch as error was only usable if I defined the error variable globally. This just sets it back.

---

If you take a look at this issue I just submitted, https://github.com/storyscript/storyscript/issues/1094

`as error` was working perfectly for me not too long ago, with the try catch implementation on the runtime of course. Trying to debug it, I realized I could simply define the variable globally to get the runtime to compile. When passing this to the context, I realized I was deleting it after the catch block executes. In the end, removing the global variable I defined. It's a bit of a weird situation, but the above is more or less a temporary fix to that problem.

I'm not sure if the compiler should throw an error since the variable was already defined, but I am experiencing an issue a bit intertwined in this.

---

> Is there a reason why we would not want to define variables to not be accessible after execution? 

I was talking about the `as error`.
Though there are many other example, like `foreach` or `while`

> Generally, a catch block would be able to define, and access globals variables.

Definitely a big no on defining _global_ variables. You shouldn't be able to ever add new global variables. If you're talking about the parent scope, that's a no too.
Every block (indent) creates a new context with the following rules:
- you can access to the parent scope variables
- you can write to the parent scope variables (if they are not readonly)
- you can't define a variable from the parent scope (that would make things ambiguous, you'll get an error)
- the variables that you define in your scope are gone after you leave the scope

The idea here is to force the user to be very explicit about his code.

```coffee
c = 0 # we made this explicit, it won't work otherwise
foreach arr as e
  c += e
```

The same applies to try/catch. That being said there's on exception: `if` statements. If you define a variable in all statements, then it will be hoisted to the parent scope. However, if you define it in only the if-block, but not the else-block, you'll get a nice compiler error (though that last part isn't fully implemented yet - see https://github.com/storyscript/storyscript/issues/1080 for details). We could apply a similar logic to `try/catch`.

> In this specific case, I actually added the code due to a recent error I discovered in the compiler. 

Please don't try to workaround things unnecessarily. Ask e.g. on #storyscript as we very likely can fix things for you.
This just creates more tech debt. In this case though (I assume you're talking about https://github.com/storyscript/storyscript/issues/1094) it's only partially a compiler error as the code is incorrect according to your agreement on the team summit (=https://github.com/storyscript/storyscript/issues/954). Though of course it should give a better error message and say that `catch as error` is not allowed. Anyhow, let's continue that discussion on the respective issue.

---

> We could apply a similar logic to try/catch.

-> https://github.com/storyscript/storyscript/issues/1095

---

@wilzbach, unfortunately, I wasn't able to be a part of the majority of the discussions at the team summit since I was not there. That clarifies the majority of what I'm trying to figure out. It looks like this would now need an update to the compiler in order to continue implementing the try-catch. As it is right now, the compiler enforces the syntax ```catch as error```, but it throws an error when code within the catch block tries accessing it. I'm going to revert it to exactly what I had before.

---

Trying the following code:

```
try
     throw "error"
catch
    log info msg: "Error Happened"
finally
     log info msg: "Finally"
```

Will result in
```
Error: syntax error in story at line 3, column 6

3|    catch
           ^

E0053: Unexpected end of line. Expected: ['_AS'].
```

---

> I wasn't able to be a part of the majority of the discussions at the team summit since I was not there. 

Yeah we should have kept better notes, but I tried to move all important points to GH. In this case it's https://github.com/storyscript/storyscript/issues/954.

> hat clarifies the majority of what I'm trying to figure out. It looks like this would now need an update to the compiler in order to continue implementing the try-catch. As it is right now, the compiler enforces the syntax catch as error

Sorry, yes, we apparently need to talk more.

> As it is right now, the compiler enforces the syntax catch as error

Ok -> https://github.com/storyscript/storyscript/issues/1096